array_api_tests/test_creation_functions.py

```diff
@@ -177,54 +217,75 @@ def test_linspace(start, stop, num, dtype, endpoint):
         # for i in range(1, num):
         #     assert all(equal(a[i], full((), i*(stop - start)/n + start, dtype=dtype))), f"linspace() produced an array with an incorrect value at index {i}"
 
-@given(shapes, one_of(none(), dtypes))
-def test_ones(shape, dtype):
-    kwargs = {} if dtype is None else {'dtype': dtype}
-    if dtype is None or is_float_dtype(dtype):
-        ONE = 1.0
+
+def make_one(dtype):
+    if kwargs is None or is_float_dtype(dtype):
+        return 1.0
     elif is_integer_dtype(dtype):
-        ONE = 1
+        return 1
     else:
-        ONE = True
+        return True
 
-    a = ones(shape, **kwargs)
 
-    if dtype is None:
-        # TODO: Should it actually match the fill_value?
-        # assert a.dtype in _floating_dtypes, "eye() should produce an array with the default floating point dtype"
-        pass
+@given(shapes, kwargs(dtype=none() | xps.scalar_dtypes()))
+def test_ones(shape, kw):
+    out = ones(shape, **kw)
+    dtype = kw.get("dtype", None) or float32
+    if kw.get("dtype", None) is None:
+        assert is_float_dtype(out.dtype), "ones() returned an array with dtype {x.dtype}, but should be the default float dtype"
     else:
-        assert a.dtype == dtype
-
-    assert a.shape == shape, "ones() produced an array with incorrect shape"
-    assert all(equal(a, full((), ONE, **kwargs))), "ones() array did not equal 1"
+        assert out.dtype == dtype, f"{dtype=!s}, but ones() returned an array with dtype {out.dtype}"
+    assert out.shape == shape, "ones() produced an array with incorrect shape"
+    assert all(equal(out, full((), make_one(dtype), dtype=dtype))), "ones() array did not equal 1"
```

Comment:
You used asarray() above but full() everywhere else. We're actually not very consistent about this in other places in the test suite, but it might be a good idea to pick only one. We want to keep the dependency graph of the tests simpler when possible, so that people can get meaningful test results if they've only implemented a subset of the API. It's not always possible, but it's something to strive for (and even the present tests don't do this as well as it could be).

Reply:
Thanks for explaining, makes sense. `asarray()` seems like easier to use, so I think I prefer that as the defacto standard for these kind of assertions.